# Iowa House Prices (EDA, Feature Engineering and Data Cleaning)

In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import Imputer

In [2]:
df_train = pd.read_csv('C:/Users/smhirech/Desktop/Datasets/iowa/train.csv', sep = ',')

In [3]:
df_train.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726   
std       1.112799    30.202904     20.645407   181.066207   456.098091   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000   

           ...         WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  \
count      ...        1460.000000  1460.000000    1460.000000  1460.000000   
mean       ...          94.244521    46.660274      21.954110     3.409589   
std        ...         125.338794    66.256028      61.119149    29.317331   
min        ...           0.000000     0.000000       0.000000     0.000000   
25%        ...           0.000000     0.000000       0.000000     0.000000   
50%        ...           0.000000    25.000000       0.000000     0.000000   
75%        ...         168.000000    68.000000       0.000000     0.000000   
max        ...         857.000000   547.000000     552.000000   508.000000   

       ScreenPorch     PoolArea       MiscVal       MoSold       YrSold  \
count  1460.000000  1460.000000   1460.000000  1460.000000  1460.000000   
mean     15.060959     2.758904     43.489041     6.321918  2007.815753   
std      55.757415    40.177307    496.123024     2.703626     1.328095   
min       0.000000     0.000000      0.000000     1.000000  2006.000000   
25%       0.000000     0.000000      0.000000     5.000000  2007.000000   
50%       0.000000     0.000000      0.000000     6.000000  2008.000000   
75%       0.000000     0.000000      0.000000     8.000000  2009.000000   
max     480.000000   738.000000  15500.000000    12.000000  2010.000000   

           SalePrice  
count    1460.000000  
mean   180921.195890  
std     79442.502883  
min     34900.000000  
25%    129975.000000  
50%    163000.000000  
75%    214000.000000  
max    755000.000000  

[8 rows x 38 columns]

In [4]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
Id               1460 non-null int64
MSSubClass       1460 non-null int64
MSZoning         1460 non-null object
LotFrontage      1201 non-null float64
LotArea          1460 non-null int64
Street           1460 non-null object
Alley            91 non-null object
LotShape         1460 non-null object
LandContour      1460 non-null object
Utilities        1460 non-null object
LotConfig        1460 non-null object
LandSlope        1460 non-null object
Neighborhood     1460 non-null object
Condition1       1460 non-null object
Condition2       1460 non-null object
BldgType         1460 non-null object
HouseStyle       1460 non-null object
OverallQual      1460 non-null int64
OverallCond      1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
RoofStyle        1460 non-null object
RoofMatl         1460 non-null object
Exterior1st      1460 non-n

In [5]:
df_train.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities    ...     PoolArea PoolQC Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  

[5 rows x 81 columns]

## Missing values

### Existence of missing values

In [6]:
df_train.isnull().any().any()

True

There are missing values in our dataset. We will have to handle those missing values to have a complete dataset, which is in particular mandatory for feature engineering. <br>
Let us now find how many variables contain missing values. 

In [7]:
df_na = pd.DataFrame(df_train.isnull().sum())
df_na.columns = ['var_na']

In [8]:
print("There are",len(df_na.loc[df_na['var_na'] != 0].index),"variables that contain missing values")

There are 19 variables that contain missing values


### Dropping variables with a considerable amount of NaN 

Let us find the percentage of missing values in the whole dataset for these variables. It will be important to know this information since it will influence our way of proceeding with missing values.

In [9]:
imp_todealwith = df_na.loc[df_na['var_na'] != 0]*100/1459

In [10]:
imp_todealwith

var_na
LotFrontage   17.751885
Alley         93.831391
MasVnrType     0.548321
MasVnrArea     0.548321
BsmtQual       2.535984
BsmtCond       2.535984
BsmtExposure   2.604524
BsmtFinType1   2.535984
BsmtFinType2   2.604524
Electrical     0.068540
FireplaceQu   47.292666
GarageType     5.551748
GarageYrBlt    5.551748
GarageFinish   5.551748
GarageQual     5.551748
GarageCond     5.551748
PoolQC        99.588759
Fence         80.808773
MiscFeature   96.367375

Let us consider that more than 25% of missing values is "considerable". Let us drop these columns from our table and create a new version of our training set. 

In [11]:
sup_na = (df_na.loc[df_na['var_na'] != 0]*100/1459 > 25)

In [12]:
na_drop = sup_na[sup_na['var_na'] == True]

In [13]:
na_drop.index

Index(['Alley', 'FireplaceQu', 'PoolQC', 'Fence', 'MiscFeature'], dtype='object')

So, we will drop the variables "Alley", "FireplaceQu", "PoolQC", "Fence" and "MiscFeature". 

In [14]:
df_train_v1 = df_train.drop(na_drop.index, axis = 1)

In [15]:
df_train_v1.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0   1          60       RL         65.0     8450   Pave      Reg         Lvl   
1   2          20       RL         80.0     9600   Pave      Reg         Lvl   
2   3          60       RL         68.0    11250   Pave      IR1         Lvl   
3   4          70       RL         60.0     9550   Pave      IR1         Lvl   
4   5          60       RL         84.0    14260   Pave      IR1         Lvl   

  Utilities LotConfig    ...     EnclosedPorch 3SsnPorch ScreenPorch PoolArea  \
0    AllPub    Inside    ...                 0         0           0        0   
1    AllPub       FR2    ...                 0         0           0        0   
2    AllPub    Inside    ...                 0         0           0        0   
3    AllPub    Corner    ...               272         0           0        0   
4    AllPub       FR2    ...                 0         0           0        0   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
0       0      2    2008        WD         Normal     208500  
1       0      5    2007        WD         Normal     181500  
2       0      9    2008        WD         Normal     223500  
3       0      2    2006        WD        Abnorml     140000  
4       0     12    2008        WD         Normal     250000  

[5 rows x 76 columns]

In [16]:
imp_todealwith = imp_todealwith.drop(na_drop.index)

### Simple analysis of the variables with missing values

In [17]:
imp_todealwith

var_na
LotFrontage   17.751885
MasVnrType     0.548321
MasVnrArea     0.548321
BsmtQual       2.535984
BsmtCond       2.535984
BsmtExposure   2.604524
BsmtFinType1   2.535984
BsmtFinType2   2.604524
Electrical     0.068540
GarageType     5.551748
GarageYrBlt    5.551748
GarageFinish   5.551748
GarageQual     5.551748
GarageCond     5.551748

Quantitative variables: 
* __MasVnrArea__: Masonry veneer area in square feet

Qualitative variables: 
* __MasVnrType__: Masonry veneer type (Brick Common, Brick Face, Cinder Block, Stone, None) => NaN real missing values
* __BsmtQual__: Evaluates the height of the basement (Excellent, Good, Typical, Fair, Poor, NaN for No Basement) => NaN NOT real missing values
* __BsmtCond__: Evaluates the general condition of the basement (Excellent, Good, Typical, Fair, Poor, NaN for No Basement) => NaN NOT real missing values
* __BsmtExposure__ : Refers to walkout or garden level walls (Good, Average, Minimum, No Exposure, NaN for No Basement) => NaN are NOT real missing values + why are they more NaN than in BsmtQual and BsmtCond ? 
* __BsmtFinType1__ : Rating of basement finished area
* __BsmtFinType2__ : Rating of basement finished area (if multiple types) (Good Living Quarters, Average Living Quarters, Below Average Living Quarters, Average Rec Room, Low Quality, Unfinshed, NaN for No Basement)
* __Electrical__ : 
* __GarageType__ : 
* __GarageYrBlt__ : 
* __GarageFinish__ : 
* __GarageQual__ :
* __GarageCond__ : 



#### Dealing with Masonry variables

In [18]:
df_train_v1[df_train_v1['MasVnrType'].isnull()]

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape  \
234    235          60       RL          NaN     7851   Pave      Reg   
529    530          20       RL          NaN    32668   Pave      IR1   
650    651          60       FV         65.0     8125   Pave      Reg   
936    937          20       RL         67.0    10083   Pave      Reg   
973    974          20       FV         95.0    11639   Pave      Reg   
977    978         120       FV         35.0     4274   Pave      IR1   
1243  1244          20       RL        107.0    13891   Pave      Reg   
1278  1279          60       RL         75.0     9473   Pave      Reg   

     LandContour Utilities LotConfig    ...     EnclosedPorch 3SsnPorch  \
234          Lvl    AllPub    Inside    ...                 0         0   
529          Lvl    AllPub   CulDSac    ...               200         0   
650          Lvl    AllPub    Inside    ...                 0         0   
936          Lvl    AllPub    Inside    ...                 0         0   
973          Lvl    AllPub    Corner    ...                 0         0   
977          Lvl    AllPub    Inside    ...                 0         0   
1243         Lvl    AllPub    Inside    ...                 0         0   
1278         Lvl    AllPub    Inside    ...                 0         0   

     ScreenPorch PoolArea MiscVal MoSold  YrSold  SaleType  SaleCondition  \
234            0        0       0      5    2010        WD         Normal   
529            0        0       0      3    2007        WD         Alloca   
650            0        0       0      5    2008        WD         Normal   
936            0        0       0      8    2009        WD         Normal   
973            0        0       0     12    2008       New        Partial   
977            0        0       0     11    2007       New        Partial   
1243           0        0       0      9    2006       New        Partial   
1278           0        0       0      3    2008        WD         Normal   

      SalePrice  
234      216500  
529      200624  
650      205950  
936      184900  
973      182000  
977      199900  
1243     465000  
1278     237000  

[8 rows x 76 columns]

There are 8 rows with 

#### Dealing with Basement variables

In [19]:
pbBsmt = df_train_v1[df_train_v1['BsmtExposure'].isnull()]['BsmtCond']

In [20]:
pbBsmt.dropna()

948    TA
Name: BsmtCond, dtype: object

The $948^{th}$ row seems to have an input error in it since we have: 
* BsmtExposure = "No Basement" 
* BsmtCondition = TA for "Typical" and not "No Basement"

In [21]:
pbBsmt2 = df_train_v1[df_train_v1['BsmtFinType2'].isnull()]['BsmtCond']

In [22]:
pbBsmt2.dropna()

332    TA
Name: BsmtCond, dtype: object

The $332^{th}$ row seems to have an input error in it since we have: 
* BsmtExposure = "No Basement" 
* BsmtCondition = TA for "Typical" and not "No Basement"

### Script for imputation by the mean of example (LotFrontage)

In [23]:
imp = Imputer(missing_values = "NaN", strategy = "mean", axis = 0)

In [24]:
q = imp.fit_transform(df_train_v1['LotFrontage'].reshape(-1,1))

C:\Users\smhirech\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead
  """Entry point for launching an IPython kernel.


In [25]:
df_train_v1['LotFrontage'] = q 

In [26]:
#imp_todealwith = imp_todealwith.drop(['LotFrontage'])

### Script Imputation by PCA on test example

Maybe later we will use impution by PCA for quantitative variables, so let us write a script with a test DataFrame. The test DataFrame is composed of two columns with 50 observations each. The first column has 10 missing values. We want to use imputation by PCA, that is to say: 
* First, we impute by the mean 
* We perform a PCA by keeping all the components
* We extract the new values of the former missing values 
* We take the original dataset and replace the missing values by these new values.
* We iterate the last three bullet points till convergence

In [27]:
from sklearn.utils.validation import check_array, check_is_fitted, check_X_y
from sklearn.decomposition import PCA
import random

In [28]:
nb_iter = 10

In [29]:
X = np.random.normal(loc = 50, scale = 2, size = 50)
Y = np.random.normal(loc = 20, scale = 10, size = 50)

In [30]:
test = pd.DataFrame({'X': X, 'Y': Y})

In [31]:
add_na = random.sample(range(0,49),10)
add_na

[33, 8, 34, 23, 47, 5, 43, 13, 31, 39]

In [32]:
test.iloc[add_na,0] = np.nan

In [33]:
test.head()

X          Y
0  52.766620  18.243092
1  46.101267  16.318711
2  49.605578  30.610559
3  48.131038  33.571101
4  50.523598  14.147987

In [34]:
test = check_array(test, dtype=np.float64, force_all_finite=False)
test_nan = np.isnan(test)
most_by_nan = test_nan.sum(axis=0).argsort()[::-1]
imputed = Imputer(strategy='mean').fit_transform(test)
new_imputed = imputed.copy()
estimators_ = [PCA(n_components=int(np.sqrt(min(test.shape))), whiten=True)]
statistics_ = np.ma.getdata(test)
gamma_ = []

In [35]:
for iter in range(nb_iter):
    if len(estimators_) > 1:
        for i in most_by_nan:

            test_s = np.delete(new_imputed, i, 1)
            y_nan = test_nan[:, i]

            test_train = test_s[~y_nan]
            y_train = new_imputed[~y_nan, i]
            test_unk = test_s[y_nan]

            estimator_ = estimators_[i]
            estimator_.fit(test_train, y_train)
            if len(test_unk) > 0:
                new_imputed[y_nan, i] = estimator_.predict(test_unk)

    else:
        estimator_ = estimators_[0]
        estimator_.fit(new_imputed)
        new_imputed[test_nan] = estimator_.inverse_transform(estimator_.transform(new_imputed))[test_nan]

    gamma = ((new_imputed-imputed)**2/(1e-6+new_imputed.var(axis=0))).sum()/(1e-6+test_nan.sum())
    gamma_.append(gamma)
    if np.abs(np.diff(gamma_[-2:])) < 1e-3:
        break

C:\Users\smhirech\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:24: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.


In [36]:
test = pd.DataFrame({'X': X, 'Y': Y})
test.iloc[add_na,0] = np.nan

In [37]:
test = check_array(test, copy=True, dtype=np.float64, force_all_finite=False)
test_nan = np.isnan(test)
imputed = Imputer(strategy='mean').fit_transform(test)

In [38]:
if len(estimators_) > 1:
    for i, estimator_ in enumerate(estimators_):
        test_s = np.delete(imputed, i, 1)
        y_nan = test_nan[:, i]

        test_unk = test_s[y_nan]
        if len(test_unk) > 0:
            test[y_nan, i] = estimator_.predict(test_unk)

else:
    estimator_ = estimators_[0]
    test[test_nan] = estimator_.inverse_transform(estimator_.transform(imputed))[test_nan]

In [39]:
test

array([[ 5.27666196e+01,  1.82430925e+01],
       [ 4.61012667e+01,  1.63187109e+01],
       [ 4.96055779e+01,  3.06105591e+01],
       [ 4.81310378e+01,  3.35711010e+01],
       [ 5.05235983e+01,  1.41479867e+01],
       [ 4.97019062e+01,  3.15687601e+01],
       [ 5.07138988e+01,  1.84819654e+01],
       [ 5.18111687e+01,  1.34320265e+01],
       [ 4.95716086e+01,  2.58799916e+01],
       [ 5.49725803e+01,  2.11680544e+01],
       [ 4.94589913e+01,  1.90132329e+01],
       [ 4.77824670e+01,  2.53738450e+01],
       [ 5.15783905e+01,  3.08107970e+01],
       [ 4.94169999e+01,  1.91298051e+01],
       [ 5.16440262e+01,  1.48068084e+01],
       [ 4.90549173e+01,  1.32297329e+01],
       [ 5.16709359e+01,  3.30691523e+01],
       [ 4.62533271e+01,  1.98453213e+01],
       [ 4.80063505e+01,  3.08299004e+01],
       [ 4.94503892e+01,  2.71329414e+01],
       [ 4.81509978e+01,  1.60707189e+01],
       [ 4.95766407e+01,  6.70483467e+00],
       [ 5.13304662e+01,  1.93924211e+01],
       [ 4.